### ISSUE TITLE
Async futures created using MasterReplica connections in a transaction context aren't completing

### ISSUE THREAD

**Issue by vaietc** [lettuce-io/lettuce-core#1729]
## Bug Report

*First of all: Have you checked the docs https://github.com/lettuce-io/lettuce-core/wiki, GitHub issues, or Stack Overflow whether someone else has already reported your issue?*

Yes, did not find another issue similar to this. 

#### Current Behavior

I'm upgrading an old system using Lettuce 4.4 to Lettuce 6 (on release 6.0.1 currently) - our production Redis nodes running Redis 6 are in AWS Elasticache (however, the behaviour described here is reproducible on a standalone Redis server as well). We leverage the [MasterReplica API](https://lettuce.io/core/release/reference/#readfrom.masterreplica-connections) to create connections and also leverage the transactions API for pipelining requests. Roughly, this looks like the following on Redis:

```
MULTI
HMSET key1 ...
HMSET key2 ...
EXEC
```
Each of the above commands map to Lettuce connection.async() calls which all return a Future. We wait on all the Futures to complete before proceeding with the rest of our application code. We've found with Lettuce 6 that the transaction command Futures (here, the HMSETs) do not complete.. even though the EXEC Future completes just fine and we can see the transaction committed on the server. This leads to us incorrectly assuming we Timed out committing the transaction even though everything was fine on the server.

#### Input Code

<details>
Added an integration test that shows the behaviour

<summary>Input Code</summary>

```java
@Test
public void testBadFutures() throws Exception {
        StatefulRedisMasterReplicaConnection<String, String> connection = MasterReplica.connect(RedisClient.create(),
                StringCodec.UTF8, getNodeList());
        connection.setReadFrom(ReadFrom.UPSTREAM);
        RedisFuture<String> future1 = connection.async().multi();
        RedisFuture<String> future2 = connection.async().hmset("testkey1",
                CollectionUtil.hashMapOf("k1", "v1", "k2", "v2"));
        RedisFuture<String> future3 = connection.async().hmset("testkey2",
                CollectionUtil.hashMapOf("k1", "v1", "k2", "v2"));
        RedisFuture<TransactionResult> future4 = connection.async().exec();
        connection.flushCommands();
        if(!LettuceFutures.awaitAll(5, TimeUnit.SECONDS, Arrays.asList(future1, future2, future3, future4)
                .toArray(new RedisFuture[4]))) {
            System.out.println("Timed out!"); // This happens
        }
        // Lets examine individual futures. 
        System.out.println("Got " + future1.get() + " from multi"); // OK
        System.out.println("Got " + future4.get() + " from exec"); // OK and has results of the two HMSets (both OKs)
        System.out.println("Got " + future2.get(1, TimeUnit.SECONDS) + " from hmset"); // Times out and crashes
        System.out.println("Got " + future3.get(1, TimeUnit.SECONDS) + " from hmset"); // Same if we allow the code to get here.
} 

/**
* These nodes are forwarded ports from Elasticache nodes running in my dev AWS account. All nodes are verified reachable using redis-cli. 
*/
private List<RedisURI> getNodeList() {
        return Arrays.asList(RedisURI.builder()
                .withHost("127.0.0.1")
                .withPort(6379)
                .withSsl(true)
                .withVerifyPeer(false)
                .build(),
                RedisURI.builder()
                        .withHost("127.0.0.1")
                        .withPort(6380)
                        .withSsl(true)
                        .withVerifyPeer(false)
                        .build(),
                RedisURI.builder()
                        .withHost("127.0.0.1")
                        .withPort(6381)
                        .withSsl(true)
                        .withVerifyPeer(false)
                        .build());
    }
```

</details>

#### Expected behavior/code

All Futures created by this connection should complete when the commands complete. Moreover, when using a simple Redis client (not the MasterSlave API) and connecting to the Primary node, the same flow works just fine with all Futures completing. 

#### Environment

- Lettuce version(s): 6.0.1
- Redis version: 6.0.5 (Elasticache), 6.2.1 (local machine)

#### Possible Solution

No solution yet but when digging through, we noted that the [TransactionalCommand](https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/protocol/TransactionalCommand.java#L33) instance created for each command inside the MULTI..EXEC block has a count of '2' which per the docs makes sense as the future is completed once for the command and once for exec. However, when debugging, we notice that these futures have a count of 1 (not 0 as expected) after execution of the exec indicating that at least one of the listeners is not completing correclty? Hopeully, that gives a clue but we haven't figured out exactly what's wrong. 

#### Additional context

Couple of questions from us other than the bug report:

* Is it safe to rely on just the output of the Exec future alone since it already contains the output from individual commands in the transaction?
* Is the usage pattern of MasterReplica connections in async transactions erroneous or weird in some way?

**Comment by vaietc** [comment]
Hello, just checking in on this issue.. while we wait for triage, can you help answer this question: 

`Is it safe to rely on just the output of the Exec future alone since it already contains the output from individual commands in the transaction?`

Thanks!

**Comment by mp911de** [comment]
Transactions and clustered/replicated connections do not work well together. Some commands might hit other servers and Redis transactions require to remain on a single node. Therefore, if you want to make use of transactions, connect to the actual node that should host the transaction.

There isn't really something that we could do here as we do not know the intent of the transaction up front, we cannot pin the transaction to a particular node.

**Comment by vaietc** [comment]
> Transactions and clustered/replicated connections do not work well together. Some commands might hit other servers and Redis transactions require to remain on a single node.

I agree with this statement in general but I'd like to provide some context on our production scenario. We're using a single primary, two replica set up with all writes directed at the single primary. We like to use MasterReplica connections instead of writing to the dedicated primary to allow for seamless primary failover management. The failure scenario you described should be rare for us given primary failovers aren't all that common and we see the described issue all the time. 

Given this, I do believe this is still a bug?

**Comment by mp911de** [comment]
It would make sense to investigate what happens under the hood and where commands are sent to. As per the command flags, `EXEC` and `MULTI` aren't considered read-only commands so your sequence of commands should actually work.

I'm wondering whether it would make sense for clustered API's to extend the transaction API to specify the intent of the operation so that a transaction can be pinned to a specific node to make it work.
